File: sdks/php/src/CreditNotes/Requests/CreateCreditNoteDto.php
<?php

namespace NovaBilling\CreditNotes\Requests;

use NovaBilling\Core\Json\JsonSerializableType;
use NovaBilling\Core\Json\JsonProperty;
use NovaBilling\CreditNotes\Types\CreateCreditNoteDtoReason;
use NovaBilling\Core\Types\ArrayType;
use NovaBilling\CreditNotes\Types\CreateCreditNoteDtoStatus;

class CreateCreditNoteDto extends JsonSerializableType
{
    /**
     * @var string $invoiceId Invoice ID to credit against
     */
    #[JsonProperty('invoiceId')]
    public string $invoiceId;

    /**
     * @var string $customerId Customer ID
     */
    #[JsonProperty('customerId')]
    public string $customerId;

    /**
     * @var float $amount Credit amount
     */
    #[JsonProperty('amount')]
    public float $amount;

    /**
     * @var string $currency Currency
     */
    #[JsonProperty('currency')]
    public string $currency;

    /**
     * @var value-of<CreateCreditNoteDtoReason> $reason
     */
    #[JsonProperty('reason')]
    public string $reason;

    /**
     * @var ?array<string, mixed> $metadata Additional metadata
     */
    #[JsonProperty('metadata'), ArrayType(['string' => 'mixed'])]
    public ?array $metadata;

    /**
     * @var ?value-of<CreateCreditNoteDtoStatus> $status Override status for imports
     */
    #[JsonProperty('status')]
    public ?string $status;

    /**
     * @var ?string $createdAt Backdate createdAt (ISO 8601). For data imports.
     */
    #[JsonProperty('createdAt')]
    public ?string $createdAt;

    /**
     * @param array{
     *   invoiceId: string,
     *   customerId: string,
     *   amount: float,
     *   currency: string,
     *   reason: value-of<CreateCreditNoteDtoReason>,
     *   metadata?: ?array<string, mixed>,
     *   status?: ?value-of<CreateCreditNoteDtoStatus>,
     *   createdAt?: ?string,
     * } $values
     */
    public function __construct(
        array $values,
    ) {
        $this->invoiceId = $values['invoiceId'];
        $this->customerId = $values['customerId'];
        $this->amount = $values['amount'];
        $this->currency = $values['currency'];
        $this->reason = $values['reason'];
        $this->metadata = $values['metadata'] ?? null;
        $this->status = $values['status'] ?? null;
        $this->createdAt = $values['createdAt'] ?? null;
    }
}

File: sdks/csharp/src/NovaBilling/CreditNotes/Requests/CreateCreditNoteDto.cs
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record CreateCreditNoteDto
{
    /// <summary>
    /// Invoice ID to credit against
    /// </summary>
    [JsonPropertyName("invoiceId")]
    public required string InvoiceId { get; set; }

    /// <summary>
    /// Customer ID
    /// </summary>
    [JsonPropertyName("customerId")]
    public required string CustomerId { get; set; }

    /// <summary>
    /// Credit amount
    /// </summary>
    [JsonPropertyName("amount")]
    public required double Amount { get; set; }

    /// <summary>
    /// Currency
    /// </summary>
    [JsonPropertyName("currency")]
    public required string Currency { get; set; }

    [JsonPropertyName("reason")]
    public required CreateCreditNoteDtoReason Reason { get; set; }

    /// <summary>
    /// Additional metadata
    /// </summary>
    [JsonPropertyName("metadata")]
    public Dictionary<string, object?>? Metadata { get; set; }

    /// <summary>
    /// Override status for imports
    /// </summary>
    [JsonPropertyName("status")]
    public CreateCreditNoteDtoStatus? Status { get; set; }

    /// <summary>
    /// Backdate createdAt (ISO 8601). For data imports.
    /// </summary>
    [JsonPropertyName("createdAt")]
    public string? CreatedAt { get; set; }

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
